### ISSUE TITLE
Hung in infinite loop

### ISSUE THREAD

**Issue by scinos** [scinos/webpack-plugin-hash-output#16]
The fix for #15 likely introduced an infinite loop. Builds froze with:

> 93% after asset optimization Update chunks

**Comment by scinos** [comment]
A bit of background: in order to rename the files, first we have to find which file references which. If a file is not referenced by anybody, we can hash it and rename it without problems. But if a file A references a file B, we first have to process B before we can hash+rename A.

To do this, the plugin uses Webpack's chunk graph to discover all files and sort them based on the number of references to other files it contains, so then we can hash+rename each file in sequence.

But because Webpack allows circular references for async files, I had to change the algorithm to account for them (fix for #15). But I introduced a bug where that algorithm might end up in an infinite loop (this bug). Is hard to test because I don't know which exact structure produces that infinite loop :(

What I did is rework the algorithm for scratch, using the new ChunkGroup object in Webpack 4. I ended up with a smaller, safer and faster algorithm that passes all existing tests. It _should not_ produce an infinite loop, but because Webpack's Chunk graph is not super well documented, it is hard to guarantee.

@salzhrani @fi5u @mnpenner Would you be able to test that new algorithm in your projects? I pushed it to the branch `candidate-fix-16`. If it works, I'll merge it and release a new version of the plugin. If it doesn't, back to the drawing board.

**Comment by mnpenner** [comment]
Installed with

```bash
yarn add 'https://github.com/scinos/webpack-plugin-hash-output.git#candidate-fix-16' --dev
```

Looks like it works!! Finally, my CSS and JS can be cache-busted individually, and all my chunks don't get new hashes when code-splitting a new chunk. Woohoo!

Looks like it costs 2 or 3 seconds of build time (37 vs 40), but I only need it for production so that's not bad. Coolbeans.

Thank you @scinos !

**Comment by scinos** [comment]
Thanks for testing @mnpenner @fi5u , you rock!

I released v3.0.0 with the fix.
